components/script/script_module.rs

```diff
@@ -0,0 +1,1313 @@
+                        ))),
+                        Some(ModuleHandler::new(Box::new(
+                            task!(all_failure_reject: move || {
+                                println!("promise all failed");
+                                let promise = reject_promise.root();
+                                promise.reject_native(&());
+                            }),
+                        ))),
+                    );
+
+                    promise_all.append_native_handler(&handler);
+
+                    return Some(promise_all);
+                }
+            }
+        },
+        Err(err) => {
+            module_tree.set_error(Some(err));
+        },
+    }
+
+    None
+}
+
+#[allow(unsafe_code)]
+/// https://html.spec.whatwg.org/multipage/#fetch-the-descendants-of-a-module-script
+fn fetch_module_descendants(
+    owner: &ModuleOwner,
+    module_tree: &ModuleTree,
+    destination: Destination,
+    mut visited: HashSet<ServoUrl>,
+) -> Result<Vec<Rc<Promise>>, ModuleError> {
+    println!("Start to load dependencies of {}", module_tree.url.clone());
+
+    let global = owner.global();
+
+    module_tree.set_status(ModuleStatus::FetchingDescendants);
+
+    module_tree
+        .resolve_requested_modules(&global, &mut visited)
+        .map(|requested_urls| {
+            module_tree.append_descendant_urls(requested_urls.clone());
+
+            requested_urls
+                .iter()
+                .map(|requested_url| {
+                    println!("Requested url: {}", requested_url);
+
+                    {
+                        let module_map = global.get_module_map().borrow();
+                        let descendant_tree = module_map.get(&requested_url.clone());
+                        if let Some(module_tree) = descendant_tree {
+                            let status = module_tree.get_status();
+                            let promise = module_tree.get_promise().borrow();
+
+                            println!("See existing descendant: {}", requested_url.clone());
+
+                            return match promise.as_ref() {
+                                Some(p) => {
+                                    println!(
+                                        "Visited array of requested url {}: it's visited {:?} and status is {:?}",
+                                        requested_url.clone(),
+                                        visited.contains(&requested_url.clone()),
+                                        status.clone()
+                                    );
+
+                                    if visited.contains(&requested_url.clone()) {
+                                        match status {
+                                            ModuleStatus::Initial | ModuleStatus::Fetching => {},
+                                            ModuleStatus::FetchingDescendants => {
+                                                let descendant_urls = module_tree.get_descendant_urls().borrow();
```

Comment:
do we need to explicitly handle this case? Won't its own promise resolve when all the descendants resolve? I'm a bit confused by this section

Reply:
I used this way to solve promises for cyclic modules 🤔 I'm still thinking if there's any better way to make it easier to understand.

Let me explain why we need to do this here:

for example,

https://github.com/servo/servo/blob/7da8d75a7e2a1f07bd09b8fb03b404ce4392a2af/tests/wpt/web-platform-tests/html/semantics/scripting-1/the-script-element/module/imports.html#L59-L62

In this test case, 

```
imports-cycle.js
└── imports-cycle-a.js
    └── imports-cycle-b.js
         └── imports-cycle-a.js
             └── imports-cycle-b.js
.....
```

`imports-cycle.js` imports `a` file and `a` file imports `b` file.
If we don't find some point to resolve the promise and just return it,  it will be a deadlock in this case.
So, current idea here is, if the module is under `FetchingDescendants` status and all of its descendants are also fetching descendants or fetched, then we will resolve or reject its promise (reject when there's any error).

Reply:
Won't this fail if the cycle has three steps (a -> b -> c -> a) though?